### ISSUE TITLE
how-to-upgrade-to-new-version: maintenance mode breaks oro:message-queue:consume

### ISSUE THREAD

**Issue by martinvonwittich** [oroinc/crm#294]
The site https://www.orocrm.com/documentation/current/dev-guide/cookbook/how-to-upgrade-to-new-version says:

> 1 - cd to the crm root folder and switch the application to the maintenance mode.
> ...
> 9 - Run the consumer(s).

Running the consumer seemed to suspiciously little; no output, no load/CPU usage in `htop`, and I wasn't even able to kill it with CTRL+C or SIGTERM. I killed it with SIGKILL and ran it again with `-vvv`:

```
orocrm.iserv.eu /var/www/orocrm # sudo -u www-data app/console oro:message-queue:consume --env prod -vvv
DEBUG: Set logger to the context {"extension":"Oro\Component\MessageQueue\Consumption\Extension\LoggerExtension"}
INFO: Start consuming
DEBUG: Switch to a queue oro.default
NOTICE: The maintenance mode has been activated. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
INFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
^CINFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
^CINFO: Waiting for the maintenance mode deactivation. {"extension":"Oro\Bundle\MessageQueueBundle\Consumption\Extension\ResettableExtensionWrapper"}
```

Seriously?

**Comment by yurio** [comment]
Hi, @martinvonwittich.

Sorry, but p 2 says: "Stop the cron tasks".

You should run maintenance mode in production environment. If you just develop some project, you can skip this step. But do not forgot to stop the consumers before update.

One of the things that do maintenance is turns off the processing of new messages by consumers to be able to process the update correctly.

After you will turn of the maintenance mode, all consumers will be turned off. It was done to start the consumers with updated code and caches.

**Comment by martinvonwittich** [comment]
> Sorry, but p 2 says: "Stop the cron tasks".

I did that ;)

> You should run maintenance mode in production environment. If you just develop some project, you can skip this step. But do not forgot to stop the consumers before update.

This is a production instance intended as an evaluation for our sales department. The reason that I posted a link to dev-guide is that the coworker from sales sent this link to me with the instruction "please upgrade our 2.4 instance to 2.5"; sorry if this caused any confusion. As far as I can tell though, the most recent non-dev upgrade guide ( https://www.orocrm.com/documentation/2.3/cookbook/how-to-upgrade-to-new-version ) doesn't seem to differ significantly.

> One of the things that do maintenance is turns off the processing of new messages by consumers to be able to process the update correctly.

Yes, that seems sensible and I understand the reasoning behind that, but it doesn't change the fact that the upgrade guide is broken. It tells you to turn on the maintenance mode (step 1), and later on it tells you to start the consumer (step 9) - without disabling the maintenance mode beforehand. But the consumer will just silently hang without doing anything because maintenance mode is enabled.

**Comment by mbessolov** [comment]
That's correct - the consumers will do nothing until you switch the maintenace mode off in step 11.

**Comment by mbessolov** [comment]
To clarify - technically, the last few steps can be performed in any order, it just happens that the upgrade guide includes them in this particular sequence. There is nothing wrong with that.

**Comment by martinvonwittich** [comment]
Ah - I think I finally understand. Until now I assumed that the `oro:message-queue:consume` call would only process any outstanding queue elements until the queue had been emptied, and would then terminate, and that this would be a step I had to complete before I could continue with the upgrade process. I had interpreted "Run the consumer" as "start the consumer and wait for its completion", and not as "start the consumer and leave it running". The fact that the documentation doesn't seem to include a step "stop the consumer" before starting the upgrade only reinforced that.

Will enabling maintenance mode automatically kill any running consumers, or will they just idle until maintenance mode has been disabled again? Maybe a hint about this should be added to the documentation to clarify whether the consumer has to be stopped manually, or why it is not necessary to do so.

It would also be nice if the step "Run the consumer" could be clarified. If the running consumer has never been stopped and won't stop on its own, then starting another one seems to be pretty pointless. If any kind of process supervision is being used (`supervisord` or `systemd`), and as far as I understand OroCRM, this is absolutely the suggested configuration, the documentation should at least contain a hint for the administrator that he might want to check if the supervisor will take care of the message consumer.
